pyxem/signals/segments.py

```diff
@@ -0,0 +1,444 @@
+            by the Li threshold method in scikit-image. If False
+            (default), the mask is the boolean VDF image.
+        exclude_border : int or True, optional
+            If non-zero integer, peaks within a distance of
+            exclude_border from the boarder will be discarded. If True,
+            peaks at or closer than min_distance of the boarder, will be
+            discarded.
+
+        References
+        ----------
+        [1] http://scikit-image.org/docs/dev/auto_examples/segmentation/
+            plot_watershed.html
+        [2] http://scikit-image.org/docs/dev/auto_examples/xx_applications/
+            plot_coins_segmentation.html#sphx-glr-auto-examples-xx-
+            applications-plot-coins-segmentation-py
+
+        Returns
+        -------
+        learning_segment : LearningSegment
+            LearningSegment where the loadings have been segmented and some
+            factors have been repeated according to the new number of loading
+            segments.
+        """
+
+        factors = self.factors.copy()
+        loadings = self.loadings.copy()
+        loadings_shape_x = loadings.data.shape[1]
+        loadings_shape_y = loadings.data.shape[2]
+        factors_shape_x = factors.data.shape[1]
+        factors_shape_y = factors.data.shape[2]
+
+        loadings_segments = np.array(loadings.map(
+            separate_watershed, show_progressbar=True, inplace=False,
+            min_distance=min_distance, min_size=min_size, max_size=max_size,
+            max_number_of_grains=max_number_of_grains,
+            marker_radius=marker_radius, threshold=threshold,
+            exclude_border=exclude_border), dtype=np.object)
+
+        segments, factors_of_segments = [], []
+        num_segs_tot = 0
+        for loading_segment, factor in zip(loadings_segments, factors):
+            segments = np.append(segments, loading_segment)
+            num_segs = np.shape(loading_segment)[0]
+            factors_of_segments = np.append(
+                factors_of_segments, np.broadcast_to(
+                    factor, (num_segs, factors_shape_x, factors_shape_y)))
+            num_segs_tot += num_segs
+        segments = segments.reshape(
+            (num_segs_tot, loadings_shape_y, loadings_shape_x))
+        factors_of_segments = factors_of_segments.reshape(
+            (num_segs_tot, factors_shape_x, factors_shape_y))
+
+        if min_intensity_threshold > 0:
+            delete_indices = list(map(
+                lambda x: x.max() < min_intensity_threshold, segments))
+            delete_indices = np.where(delete_indices)
+            segments = np.delete(segments, delete_indices, axis=0)
+            factors_of_segments = np.delete(
+                factors_of_segments, delete_indices, axis=0)
+
+        try:
+            segments = Signal2D(segments).transpose(navigation_axes=[0],
+                                                    signal_axes=[2, 1])
+        except TraitError:
```

Comment:
This condition needs a test. This also seems like a potentially brittle/unreliable way to pick up this issue?

Reply:
As on the similar one earlier, I don't want to be writing try/except for stuff like this. Will fix.